src/metabase/models/visualization_settings.cljc

```diff
@@ -23,144 +23,139 @@
-
-;; target ID can be the auto generated ID or fully qualified name for serialization
-(s/def ::link-target-id (s/or :int int? :fully-qualified-name string?))
-(s/def ::link-template string?)
-(s/def ::link-text-template string?)
-
-(s/def ::column-title string?)
-(s/def ::date-style #{"M/D/YYYY" "D/M/YYYY" "YYYY/M/D" "MMMM D, YYYY" "D MMMM, YYYY" "dddd, MMMM D, YYYY"})
-(s/def ::date-abbreviate boolean?)
-(s/def ::date-separator #{"/" "-" "."})
-(s/def ::time-style #{"HH:mm" "h:mm A" "h A"})
-(s/def ::time-enabled #{nil "minutes" "seconds" "milliseconds"})
-(s/def ::decimals pos-int?)
-(s/def ::number-separators #(or nil? (and string? (= 2 (count %)))))
-(s/def ::number-style #{"decimal" "percent" "scientific" "currency"})
-(s/def ::prefix string?)
-(s/def ::suffix string?)
-(s/def ::view-as string?)
-(s/def ::link-text string?)
-
-(s/def ::param-mapping-id string?)
-
-(s/def ::param-ref-type #{"column" "dimension" "variable" "parameter"})
-(s/def ::param-ref-id string?)
-(s/def ::param-ref-name string?)
-
-(s/def ::param-mapping-source (s/keys :req [::param-ref-id ::param-ref-type] :opt [::param-ref-name]))
-(s/def ::param-mapping-target ::param-mapping-source)
-
-(s/def ::db-column-ref (s/or :string? string? :vector? vector? :keyword? keyword?))
-
-(s/def ::entity-type #{::card ::dashboard})
+(def ^:private parameter-mapping-schema [:maybe map?])
+
+(def ^:private click-behavior-schema
+  [:map
+   [::click-behavior-type keyword?]
+   [::link-type {:optional true} :any]
+   [::parameter-mapping {:optional true} parameter-mapping-schema]
+   [::link-template {:optional true} string?]
+   [::link-text {:optional true} string?]
+   ;; target ID can be the auto generated ID or fully qualified name for serialization
+   [::link-target-id {:optional true} [:or int? string?]]])
+
+(def ^:private visualization-settings-schema
+  [:map
+   [::column-settings {:optional true} map?]
+   [::click-behavior {:optional true} click-behavior-schema]])
+
+;; (s/def ::link-target-id (s/or :int int? :fully-qualified-name string?))
+;; (s/def ::link-template string?)
+;; (s/def ::link-text-template string?)
+
+;; (s/def ::column-title string?)
+;; (s/def ::date-style #{"M/D/YYYY" "D/M/YYYY" "YYYY/M/D" "MMMM D, YYYY" "D MMMM, YYYY" "dddd, MMMM D, YYYY"})
+;; (s/def ::date-abbreviate boolean?)
+;; (s/def ::date-separator #{"/" "-" "."})
+;; (s/def ::time-style #{"HH:mm" "h:mm A" "h A"})
+;; (s/def ::time-enabled #{nil "minutes" "seconds" "milliseconds"})
+;; (s/def ::decimals pos-int?)
+;; (s/def ::number-separators #(or nil? (and string? (= 2 (count %)))))
+;; (s/def ::number-style #{"decimal" "percent" "scientific" "currency"})
+;; (s/def ::prefix string?)
+;; (s/def ::suffix string?)
+;; (s/def ::view-as string?)
+;; (s/def ::link-text string?)
+
+;; (s/def ::param-mapping-id string?)
+
+;; (s/def ::param-ref-type #{"column" "dimension" "variable" "parameter"})
+;; (s/def ::param-ref-id string?)
+;; (s/def ::param-ref-name string?)
```

Comment:
I'm not sure what to do with these. These specs are not parts of other specs and are not references by any of the fdef's. I've commented them because they look important (they describe the data that we expect) but it would be better if they validated something too.

Reply:
Some of these like `::link-text` were being used by the `::click-behavior` spec but since you removed it probably ok to remove these entirely. As far as I can tell `::number-style` was never used at all https://github.com/metabase/metabase/pull/15858/files#diff-b1e5711638b653b2a67d36d1b997a409c2625141fca2307ff3fe46c017249de9R381